## Data collection and storing:

### Fetching Data Through API In Python(automation script) and connecting data's to MySql workbench

In [1]:
#importing all the required packages
import requests
import pandas as pd
import io
#access google spreadsheets through the Google Sheets API v4
import pygsheets
import datetime as dt
#using sqlalchemy trying to establish a connection between RDS and MySql workbench
import sqlalchemy 
from sqlalchemy import create_engine

In [2]:
#Sends a GET request to data.gov.in to fetch the data from the wesite

url_data = requests.get("https://api.data.gov.in/resource/3b01bcb8-0b14-4abf-b6f2-c1bfd384ba69?api-key=579b464db66ec23bdd000001fc815e771f3a46e162d44e8e5b3ff2ab&format=csv&offset=0&limit=2000").content

In [3]:
#Reading the data which was fetched from the API as a csv file

rawData = pd.read_csv(io.StringIO(url_data.decode('utf-8')))

**So, now the data is avaialble for us to access, let's explore it. So that we can code out the script.py file in a efficient was to run on the AWS EC2 machine**

In [4]:
#This is the air_pollution data that we have fetched from the website API

rawData

id country           state       city                       station  \
0        1   India  Andhra_Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
1        2   India  Andhra_Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
2        3   India  Andhra_Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
3        4   India  Andhra_Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
4        5   India  Andhra_Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
...    ...     ...             ...        ...                           ...   
1995  1996   India   Uttar_Pradesh      Noida   Sector - 125, Noida - UPPCB   
1996  1997   India   Uttar_Pradesh      Noida   Sector - 125, Noida - UPPCB   
1997  1998   India   Uttar_Pradesh      Noida   Sector - 125, Noida - UPPCB   
1998  1999   India   Uttar_Pradesh      Noida      Sector - 62, Noida - IMD   
1999  2000   India   Uttar_Pradesh      Noida      Sector - 62, Noida - IMD   

              last_update pollutant_id  pollutant_min  pollutant_max  \
0     24-08-2022 10:00:00        PM2.5            NaN            NaN   
1     24-08-2022 10:00:00         PM10            NaN            NaN   
2     24-08-2022 10:00:00          NO2            NaN            NaN   
3     24-08-2022 10:00:00          NH3            NaN            NaN   
4     24-08-2022 10:00:00          SO2            NaN            NaN   
...                   ...          ...            ...            ...   
1995  24-08-2022 10:00:00          NH3            9.0           10.0   
1996  24-08-2022 10:00:00          SO2            1.0           12.0   
1997  24-08-2022 10:00:00           CO           43.0           52.0   
1998  24-08-2022 10:00:00        PM2.5            NaN            NaN   
1999  24-08-2022 10:00:00         PM10            NaN            NaN   

      pollutant_avg  pollutant_unit  
0               NaN             NaN  
1               NaN             NaN  
2               NaN             NaN  
3               NaN             NaN  
4               NaN             NaN  
...             ...             ...  
1995            9.0             NaN  
1996            6.0             NaN  
1997           44.0             NaN  
1998            NaN             NaN  
1999            NaN             NaN  

[2000 rows x 11 columns]

In [5]:
#Checking the size, I have limited the number of data points to 2000 because the data set will be too large to handle on my local machine.
#limit=2000 was on the url_data API call

len(rawData)

2000

In [7]:
#Checking for the information of the rawData that's collected

rawData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              2000 non-null   int64  
 1   country         2000 non-null   object 
 2   state           2000 non-null   object 
 3   city            2000 non-null   object 
 4   station         2000 non-null   object 
 5   last_update     2000 non-null   object 
 6   pollutant_id    2000 non-null   object 
 7   pollutant_min   1819 non-null   float64
 8   pollutant_max   1819 non-null   float64
 9   pollutant_avg   1819 non-null   float64
 10  pollutant_unit  0 non-null      float64
dtypes: float64(4), int64(1), object(6)
memory usage: 172.0+ KB


In [9]:
#Checking for the null values in the rawData df

rawData.isnull().sum()

id                   0
country              0
state                0
city                 0
station              0
last_update          0
pollutant_id         0
pollutant_min      181
pollutant_max      181
pollutant_avg      181
pollutant_unit    2000
dtype: int64

### Key points for data collection:

**I can see that there are 181 null values in `pollutant_min`, `pollutant_max`, `pollutant_avg` and the entire column of `pollutant_unit` have null values.** 

**NOTE: We have null values in this API call, As like I have mentioned on the Project documentation the API will be updated once in every 5 hours, So if we tend to write a script that will delete the `pollutant_unit`column from this table, it might lead to biased results as we need to consider what are the units that causes air pollution that might come up when they(data.gov.in) update the data in the future.**

In [12]:
#Extracting the data in a cleaned way
#Replacing few values in rawData in the df, as they were the few states with _ in them

rawData = rawData.replace(to_replace="Andhra_Pradesh", value ="Andhra Pradesh")
rawData = rawData.replace(to_replace ="West_Bengal", value ="West Bengal")
rawData = rawData.replace(to_replace ="Uttar_Pradesh", value ="Uttar Pradesh")
rawData = rawData.replace(to_replace ="Jammu_&_Kashmir", value ="Jammu & Kashmir")
rawData = rawData.replace(to_replace ="Arunachal_Pradesh", value ="Arunachal Pradesh")

In [13]:
#Checking for the unique values in the Sates column as we have update few values

rawData['state'].unique()

array(['Andhra Pradesh', 'Arunachal Pradesh', 'Assam', 'Bihar',
       'Chandigarh', 'Chhattisgarh', 'Delhi', 'Gujarat', 'Haryana',
       'Himachal Pradesh', 'Jammu & Kashmir', 'Karnataka', 'Kerala',
       'Madhya Pradesh', 'Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram',
       'Nagaland', 'Odisha', 'Puducherry', 'Punjab', 'Rajasthan',
       'Sikkim', 'TamilNadu', 'Telangana', 'Tripura', 'Uttar Pradesh'],
      dtype=object)

In [14]:
#Now that the values are clear, we will look for the column information

rawData.columns

Index(['id', 'country', 'state', 'city', 'station', 'last_update',
       'pollutant_id', 'pollutant_min', 'pollutant_max', 'pollutant_avg',
       'pollutant_unit'],
      dtype='object')

In [15]:
rawData.head()

id country           state       city                       station  \
0   1   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
1   2   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
2   3   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
3   4   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
4   5   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   

           last_update pollutant_id  pollutant_min  pollutant_max  \
0  24-08-2022 10:00:00        PM2.5            NaN            NaN   
1  24-08-2022 10:00:00         PM10            NaN            NaN   
2  24-08-2022 10:00:00          NO2            NaN            NaN   
3  24-08-2022 10:00:00          NH3            NaN            NaN   
4  24-08-2022 10:00:00          SO2            NaN            NaN   

   pollutant_avg  pollutant_unit  
0            NaN             NaN  
1            NaN             NaN  
2            NaN             NaN  
3            NaN             NaN  
4            NaN             NaN

**Initially, the column names appear to be appropriate for what we are trying to accomplish, but if any new features are introduced in the future, I may need to re-organize the df columns.**

In [17]:
#So, now we know that the data from API will be updated once in every 5 hours
#In order to keep a track of the time of the data being collected I need a new column with the date and time of data
#Here I have added a new column(using pd.series()) as Mesure_time to collect the time on which the data is been collected
#Using the datetime I have calculated UTC time and the timestamp on which the data is been collected

import warnings
warnings.filterwarnings('ignore')
from datetime import timedelta

rawData = rawData.append(pd.Series(), ignore_index=True)
now = dt.datetime.utcnow()+timedelta(hours=4, minutes=30)
now = now.strftime('%d-%m-%Y %H:00:00')
rawData['Measure_time'] = now

In [18]:
#Exploring the new column which we have actually collected in Measure_time column

rawData.head()

id country           state       city                       station  \
0  1.0   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
1  2.0   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
2  3.0   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
3  4.0   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   
4  5.0   India  Andhra Pradesh  Anantapur  Gulzarpet, Anantapur - APPCB   

           last_update pollutant_id  pollutant_min  pollutant_max  \
0  24-08-2022 10:00:00        PM2.5            NaN            NaN   
1  24-08-2022 10:00:00         PM10            NaN            NaN   
2  24-08-2022 10:00:00          NO2            NaN            NaN   
3  24-08-2022 10:00:00          NH3            NaN            NaN   
4  24-08-2022 10:00:00          SO2            NaN            NaN   

   pollutant_avg  pollutant_unit         Measure_time  
0            NaN             NaN  24-08-2022 11:00:00  
1            NaN             NaN  24-08-2022 11:00:00  
2            NaN             NaN  24-08-2022 11:00:00  
3            NaN             NaN  24-08-2022 11:00:00  
4            NaN             NaN  24-08-2022 11:00:00

## Connecting the AWS RDS to MySQL workbench

In [19]:
#Changing the casing of the rawData df's columns to look more slicker

rawData.columns = ['Id', 'Country', 'State', 'City', 
                'Station', 'Last_Update', 'Pollution_ID',
'Pollutant_Min', 'Pollutant_Max', 'Pollutant_Avg', 'Pollutant_Unit', 'Measure_Time']

### How this workes?
* Step 1: To create a engine with the `sqlalchemy`, which will help us to intract between Mysql workbech and the Online AWS RDS Database

* Step 2: Use the **port number, user name, db name** with the connector engine

* Step 3: Push the date that's been colled from the API to the data base "air_pollution" into the table "hourly_air_pollution"

* Step 4: Please make sure to `if_exists='append` because the data is been generated every 5 hours, we to append the data been collected to the existing data.

**`sqlalchemy` is a great tool that facilitates the communication between Python programs and databases. Most of the times, this library is used as an Object Relational Mapper (ORM) tool that translates Python classes to tables on relational databases and automatically converts function calls to SQL statements. But, here in my project I have used it as ORM since we are trying to connect to AWS RDS, but if you need a query directly from the notebook that functionality is also avaialble**

[`sqlalchemy` DOCUMENTATION](https://docs.sqlalchemy.org/en/14/)

In [20]:
#Creating the engine with your AWS PDS port number, user_name, ID and password
#'air_pollution' is the database name were I want my data to be connected and pushed into

engine = sqlalchemy.create_engine('mysql+mysqlconnector://admin:8098282280@gokultest.crip17huxagt.ap-south-1.rds.amazonaws.com:3306/air_pollution')

In [21]:
#Using the engine we have created and pushing the data into mysql workbench

rawData.to_sql(name='hourly_air_pollution', con=engine, if_exists='append', index=False)

2002

**When I created my data base in AWS RDS, I used Mysql DB as my db workbench, so I used `mysql+mysqlconnector` to connect to it; this process may differ for other RDBMS and NoSql data bases.**